### PR TITLE
Add ConnectionStatusListener to RestClient

### PR DESCRIPTION
**Pull request by PatrickTasse**
Clients can add a connection status listener to the RestClient to be
notified when the connection goes up or down. The connection status is
detected based on whether any HTTP request can be completed or not.

When a listener is added it will immediately receive the current status
through the listener interface.

Change-Id: I5f7fcbf4b94380b2ea349755a400bd61bbadc7d2
Signed-off-by: Patrick Tasse <patrick.tasse@ericsson.com>

### CODE REVIEW

**Comment by hriday-panchasara** [comment]
LGTM. Works well with the suggested implementation in https://github.com/theia-ide/theia-trace-extension/pull/671

File: tsp-typescript-client/src/protocol/rest-client.ts [hunk 1]
```diff
@@ -103,7 +109,10 @@ export class RestClient {
                 },
                 body: this.jsonStringify(body)
             });
+            const status = response.status <= 500;
```
Comment (bhufmann) [inline tsp-typescript-client/src/protocol/rest-client.ts:112]: 500 or 501 can be returned by the server, so the status should be true. You will need to be more selective when deciding that the server is down.

**Review by bhufmann — CHANGES_REQUESTED** [review]

File: tsp-typescript-client/src/protocol/rest-client.ts (hunk 1 above)
Comment (PatrickTasse) [reply tsp-typescript-client/src/protocol/rest-client.ts:112]: I looked at the status code and how we use them, and I would guess that it should be only 500 and 501 that map to true, and the others >= 502 map to false. Would that be OK?

File: tsp-typescript-client/src/protocol/rest-client.ts [hunk 2]
```diff
@@ -159,4 +168,23 @@ export class RestClient {
     protected static jsonParse(text: string): any {
         return JSONBig.parse(text);
     }
+
```
Comment (bhufmann) [inline tsp-typescript-client/src/protocol/rest-client.ts:171]: You should document what it means if connection listener is called with false or true. Probably in the ConnectionStatusListener doc

**Review by bhufmann — CHANGES_REQUESTED** [review]

**Pull request merged**